### ISSUE TITLE
support HTTP/2.0

### ISSUE THREAD

**Issue by antonikonovalov** [elastic/elasticsearch#10981]
Elastic have plane add support (or maybe plugin for) protocol HTTP/2.0? 
Like in https://github.com/grpc/grpc(used HTTP/2.0) or may be with grpc

**Comment by dakrone** [comment]
This is now possible since we have moved to Netty 4, since Netty supports HTTPP/2.0 in 4.1+ /cc @jasontedor

**Comment by jasontedor** [comment]
To be clear, Netty 4 supports HTTP/2 but there is effort required to make Elasticsearch use it, and ensure that it works properly.

**Comment by KodrAus** [comment]
Is there any new thinking on supporting h2? It seems like a fruit hanging so high it'll probably take a magic beanstalk to reach, but I'm interested in any thoughts you might have on the cost-benefit equation. I imagine there are bigger wins in first-class client support and alternative APIs that take advantage of more stream-oriented connections in supporting h2 within Elasticsearch itself over just exposing through a h2-enabled proxy.

**Comment by mantasindrasius** [comment]
Any updates on this? With the deprecated TransportClient, there should for sure be HTTP/2 or/and gRPC support.

**Comment by jasontedor** [comment]
At this time there is no update on this other than that we are not currently working on it; this not to say that we will not, only that we feel there are higher priority projects. I do not agree that the deprecation of the transport client implies that there should be HTTP/2 support. Can share why you think that the two are tied together?

**Comment by mantasindrasius** [comment]
Thanks for a quick answer.

HTTP/1.1 is limited in the sense that any concurrent operations cannot happen on the same connection and creating a new connection is a huge performance trade-off.

HTTP/2 or gRPC which is based on the former is multiplexing by nature, so is closer to the non-blocking TransportClient.

If we want to keep the transport overhead low, we can't go with REST+HTTP/1.1.

**Comment by jasontedor** [comment]
@mantasindrasius Thanks for the feedback. I agree with you on the benefits of HTTP/2.

I do think it's worth noting that with HTTP/1.1 keep alive behavior and a connection pool that you can mitigate some of the disadvantages of HTTP/1.1 versus HTTP/2.0. The best that HTTP/1.1 can offer related to multiplexing is pipelining. Yes, this suffers from head-of-line blocking but it does allow some connection reuse too.

That said, even weighing these advantages and disadvantages, I would prefer that we continue to not tie removal of the transport client to the introduction of HTTP/2. Removing the transport client will bring us major advantages in continuing to develop the server side of Elasticsearch without maintaining a client that is so intimately tied to the internal protocol. These advantages outweigh the loss of benefits of the transport client.

Rest assured that HTTP/2 is on our roadmap, just not at the top of the list at the moment.

**Comment by mantasindrasius** [comment]
Thanks for that. Do you also think there could be gRPC support for elastic or it's not on the table?

**Comment by jasontedor** [comment]
@mantasindrasius gRPC is not on the table a the moment.

**Comment by Asgoret** [comment]
@jasongoodwin ELK don't work in OKD 3.11 (Openshift) correctly because there is no support of H2\grpc protocols.
See: https://github.com/elastic/kibana/issues/7104 and https://github.com/openshift/origin-aggregated-logging/issues/1465

**Comment by ayushsangani** [comment]
Any updates on the support of Http 2.0 in ES?

**Comment by cangiremir** [comment]
Any actions to support gRPC or Http2.0?

**Comment by weitzjdevk** [comment]
Are there any news here? AWS NLB has finally ALPN support for the TLS Listener, which would help for the gRPC handshake

https://www.reddit.com/r/aws/comments/gsdo3w/nlb_now_supports_alpn_on_tls_listeners/

**Comment by bachden** [comment]
Waiting for http2 (gRPC maybe) supporting.

**Comment by bradquarry** [comment]
A team at Verizon is asking for gRPC support. +1

**Comment by YohanSciubukgian** [comment]
Do you have any update to share about HTTP/2 support please ?

**Comment by gokcesurenkok** [comment]
Do you have any plan to do that in near future ? i would like to see it if it is in your roadmap.
Thanks

**Comment by weitzj** [comment]
Any updates here?

 AWS ALB has finally gRPC support built in
GCP had it all the time

**Comment by shiyangfei** [comment]
+1 from Compass. We need this feature.

**Comment by emcpadden** [comment]
This is definitely needed!

**Comment by Farenheith** [comment]
I endorse http2 too!

One solution for me at this moment, if anyone is interested, is to avoid elasticsearch oficcial client (in a nodejs service that only uses **_search**), and to use [superagent](https://www.npmjs.com/package/superagent) + [agentkeepalive](https://www.npmjs.com/package/agentkeepalive), with, in my case of use, **maxSockets** uninformed, assuming the default value of Infinity.
This way I have the connection pool working under the hood, mitigating the http1 limitation as @jasontedor well said.

**Comment by gguillen** [comment]
+1 Upvote

**Comment by JhennerTigreros** [comment]
A team at [Platzi](https://platzi.com) is asking for gRPC support. +1

**Comment by Dyqer** [comment]
+1 Upvote

**Comment by blackheaven** [comment]
+1 Upvote

**Comment by JoeGilkey** [comment]
+1 as a nice to have.

**Comment by Jaimenms** [comment]
+1 Upvote

**Comment by cjmcgraw** [comment]
+1 this would be very helpful to my company

**Comment by doubleknd26** [comment]
+1 please

**Comment by marinrusu1997** [comment]
+1 Upvote

**Comment by KennyLindahl** [comment]
+1 Upvote

**Comment by BiditPakrashi** [comment]
Any update on GRPC support ?

**Comment by mhlee-lee** [comment]
+1 please

**Comment by fsedano** [comment]
+1 please

**Comment by nakulpathak3** [comment]
+1 Faire would like this too

**Comment by RaffysWeb** [comment]
+1 Upvote

**Comment by DaemonSnake** [comment]
+1 Upvote

**Comment by JinLiu1210** [comment]
+1 please

**Comment by vineelyalamarthy** [comment]
@clintongormley  are there discussions happening some where  like in wiki pages/ discord etc on this topic roadmap/implementation. Are open source contributors welcome to participate in those? I would love to contribute to this feature.

**Comment by lukasojd** [comment]
+1

**Comment by henningandersen** [comment]
This is not a high priority to us internally at the moment. There is no to little actual benefits from http/2 over what we have today. We would welcome a contribution, but would like to remind of our [contributions guidelines](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md), highlighting that this change is likely to require considerable test additions.

**Comment by otrosien** [comment]
@henningandersen thanks for the feedback. Out of curiosity: "There is no to little actual benefits from http/2 over what we have today" - how was that evaluated?

**Comment by loeffel-io** [comment]
+1

**Comment by magicleo** [comment]
Any update on GRPC support ?

**Comment by patpatpat123** [comment]
Upvoting this many year old idea

**Comment by firebook** [comment]
Any update? We really need it!

**Comment by DaveCTurner** [comment]
Unfortunately there's nothing in the discussion above to say _why_ this is so important. HTTP/2.0 (and now HTTP/3.0) offer significant advantages for certain applications, particularly those involving web browsers or web-browser-like interaction patterns, but we cannot see the benefits over HTTP/1.1 in the specific context of an application like ES whose clients are expected to use a pool of long-lived connections (and possibly pipelining) to achieve the same results. We just don't encounter cases where the use of HTTP/1.1 is the limiting factor for either performance or expressivity in a meaningful way right now. I'm sorry if we're missing something obvious, but without some more concrete motivation for addressing this request it's unlikely we'll take on the significant amount of work involved in doing so in the foreseeable future.

**Comment by trycoon** [comment]
@DaveCTurner HTTP/2 is of little importance to me… But I was hoping that gRPC could lower the transmission latency that JSON and REST adds to the querying. We use Elastic as a smarty cache to filter information ands answer user requests, several thousend per second. If we could shave of some additional 10ms for each 100KB JSON response by using gRPC, that would be a nice improvement.
And to be able to use gRPC we need HTTP/2… 🤗

**Comment by DaveCTurner** [comment]
> If we could shave off some additional 10ms for each 100KB JSON response by using gRPC, that would be a nice improvement.

That would indeed be very nice, but I don’t see how it could happen. The current wire protocol doesn’t add anything like that much latency.

**Comment by mselmansezgin** [comment]
@DaveCTurner the wire protocol doesn't add anything for additional latency but it also detracts from the ability to use protobuf instead of JSON. This eliminates our chance to reduce the overall size of the payload.

**Comment by DaveCTurner** [comment]
Adding HTTP/2.0 support wouldn't itself yield any meaningful savings in this area because today ES needs the data it processes to use a self-describing format such as JSON (although if you care about payload size it'd be better to use one of the other supported formats instead). Protobuf isn't naturally self-describing in the way ES needs today, so we'd either end up adding loads of metadata or else just making a thin protobuf wrapper around one of the existing formats, neither of which would be meaningfully smaller than today's wire protocol.

We could probably teach ES to speak protobuf more naturally too, separating schema from content better, but that'd be a separate enhancement. If we did that it'd be able to speak protobuf over the existing HTTP/1.1 interface as well. In other words this point is pretty much orthogonal to the conversation about HTTP/2.0.

**Comment by nakulpathak3** [comment]
+1 HTTP/2 is extremely commonly supported in 2024

**Comment by DaveCTurner** [comment]
Thanks @nakulpathak3, I think [my earlier message](https://github.com/elastic/elasticsearch/issues/10981#issuecomment-1909867176) addresses this.

**Comment by swallez** [comment]
http/2 has [ping frames](https://httpwg.org/specs/rfc7540.html#PING), which can be sent by both client and server. Having Elasticsearch support http/2 would allow it to send periodic ping frames that would keep the TCP connection alive for long-running requests by transparently informing the client application that processing is still going on.

This [blog post](https://www.honeycomb.io/blog/always-enable-keepalives) shows a concrete benefits of using them in the context of gRPC (gRPC keepalive [is just http/2 ping](https://grpc.io/docs/guides/keepalive/)).

**Comment by DaveCTurner** [comment]
A possible measurable benefit to adding HTTP/2.0 support: cheaper client cancellation.

With HTTP/1.1 a client can only cancel a request by closing the underlying TCP connection. In the steady state we expect the number of inbound client connections to be pretty constant, so closing an underlying TCP connection implies opening another one in its place. When TLS is enabled, this implies doing another TLS handshake, and that can be quite CPU-intensive.

With HTTP/2.0 the client requests are multiplexed and a client can send a message with the `RST_STREAM` flag set to cancel an individual request without needing to tear down the underlying stream.

**Comment by DaveCTurner** [comment]
ALPN is not required for connection re-use, and Elasticsearch supports (indeed, strongly expects) clients to keep using the same connection for hours/days/weeks at a time.

**Comment by DaveCTurner** [comment]
Yes, there's pretty substantial coverage of this claim in the test suite. See for instance `org.elasticsearch.http.netty4.Netty4PipeliningIT`.

It's pretty easy to verify with just `curl` too, noting e.g. the lines `ALPN: server did not agree on a protocol. Uses default.` and yet `Re-using existing connection with host 127.0.0.1`:

```
$ curl -vv --cacert '/Users/davidturner/issues/2025-08-12-tls-connection-re-use/elasticsearch-9.1.1/elastic-stack-ca.pem' --user elastic:BCgyQMtI1r9HIPkVP8xk https://127.0.0.1:9200/ https://127.0.0.1:9200/
*   Trying 127.0.0.1:9200...
* Connected to 127.0.0.1 (127.0.0.1) port 9200
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
*  CAfile: /Users/davidturner/issues/2025-08-12-tls-connection-re-use/elasticsearch-9.1.1/elastic-stack-ca.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-AES256-GCM-SHA384 / [blank] / UNDEF
* ALPN: server did not agree on a protocol. Uses default.
* Server certificate:
*  subject: CN=instance
*  start date: Aug 12 18:39:33 2025 GMT
*  expire date: Aug 11 18:39:33 2028 GMT
*  subjectAltName: host "127.0.0.1" matched cert's IP address!
*  issuer: CN=Elastic Certificate Tool Autogenerated CA
*  SSL certificate verify ok.
* using HTTP/1.x
* Server auth using Basic with user 'elastic'
> GET / HTTP/1.1
> Host: 127.0.0.1:9200
> Authorization: Basic ZWxhc3RpYzpCQ2d5UU10STFyOUhJUGtWUDh4aw==
> User-Agent: curl/8.7.1
> Accept: */*
>
* Request completely sent off
< HTTP/1.1 200 OK
< X-elastic-product: Elasticsearch
< content-type: application/json
< content-length: 530
<
{
  "name" : "node-0",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "ppik2Sy5RoGYIsQ5XG-0Vg",
  "version" : {
    "number" : "9.1.1",
    "build_flavor" : "default",
    "build_type" : "tar",
    "build_hash" : "5e94055934defa56e454868b7783b2a3b683785e",
    "build_date" : "2025-08-05T01:07:31.959947279Z",
    "build_snapshot" : false,
    "lucene_version" : "10.2.2",
    "minimum_wire_compatibility_version" : "8.19.0",
    "minimum_index_compatibility_version" : "8.0.0"
  },
  "tagline" : "You Know, for Search"
}
* Connection #0 to host 127.0.0.1 left intact
* Found bundle for host: 0x600001e08ab0 [serially]
* Can not multiplex, even if we wanted to
* Re-using existing connection with host 127.0.0.1
* Server auth using Basic with user 'elastic'
> GET / HTTP/1.1
> Host: 127.0.0.1:9200
> Authorization: Basic ZWxhc3RpYzpCQ2d5UU10STFyOUhJUGtWUDh4aw==
> User-Agent: curl/8.7.1
> Accept: */*
>
* Request completely sent off
< HTTP/1.1 200 OK
< X-elastic-product: Elasticsearch
< content-type: application/json
< content-length: 530
<
{
  "name" : "node-0",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "ppik2Sy5RoGYIsQ5XG-0Vg",
  "version" : {
    "number" : "9.1.1",
    "build_flavor" : "default",
    "build_type" : "tar",
    "build_hash" : "5e94055934defa56e454868b7783b2a3b683785e",
    "build_date" : "2025-08-05T01:07:31.959947279Z",
    "build_snapshot" : false,
    "lucene_version" : "10.2.2",
    "minimum_wire_compatibility_version" : "8.19.0",
    "minimum_index_compatibility_version" : "8.0.0"
  },
  "tagline" : "You Know, for Search"
}
* Connection #0 to host 127.0.0.1 left intact
*
```

**Comment by DaveCTurner** [comment]
Correct, that's how HTTP/1.1 works, it's nothing to do with ALPN support. I'm going to mark this sequence of messages as off-topic since they do not relate to the rest of this conversation. If you have further questions, please use the [Elasticsearch forum](https://discuss.elastic.co/c/elasticsearch) rather than continuing on this tangent here.
